File: momagic_flutter/android/src/main/kotlin/com/example/momagic_flutter/MoMagicFlutterPlugin.java
package com.example.momagic_flutter;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;


import com.momagic.DATB;
import com.momagic.PreferenceUtil;
import com.momagic.TokenReceivedListener;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

@SuppressWarnings("MoMagicFlutterPlugin")
public class MoMagicFlutterPlugin implements FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {
    @SuppressLint("StaticFieldLeak")
    static Context context;
    Activity activity;
    MethodChannel channel;
    private String  notificationToken;

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        this.activity = binding.getActivity();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
    }

    @Override
    public void onDetachedFromActivity() {
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        context = flutterPluginBinding.getApplicationContext();
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), MoMagicConstant.IZ_PLUGIN_NAME); //define the chanel name
        channel.setMethodCallHandler(this);

    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    // Handle the all methods
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case iZootoConstant.IZ_ANDROID_INIT:
                try {

                        DATB.initialize(context).setTokenReceivedListener(new TokenReceivedListener() {
                                    @Override
                                    public void onTokenReceived(String s) {
                                        Log.e("Device Token","Token"+s);

                                        channel.invokeMethod("onToken",s);
                                    }
                                })
                                .build();

                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case MoMagicConstant.IZ_SET_SUBSCRIPTION:
                try {
                    boolean setSubscription = (boolean) call.arguments;
                    DATB.setSubscription(setSubscription);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case MoMagicConstant.IZ_FIREBASE_ANALYTICS:
                try {
                    boolean trackFirebaseAnalytics = (boolean) call.arguments;
                    DATB.setFirebaseAnalytics(trackFirebaseAnalytics);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case MoMagicConstant.IZ_ADD_EVENTS:
                try {
                    String eventName = call.argument(MoMagicConstant.IZ_EVENT_NAME);
                    HashMap<String, Object> hashMapEvent = new HashMap<>();
                    hashMapEvent = call.argument(MoMagicConstant.IZ_EVENT_VALUE);
                    DATB.addEvent(eventName, hashMapEvent);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case MoMagicConstant.IZ_ADD_PROPERTIES:
                try {
                    HashMap<String, Object> hashMapUserProperty = new HashMap<>();
                    hashMapUserProperty = (HashMap<String, Object>) call.arguments;
                    DATB.addUserProperty(hashMapUserProperty);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case MoMagicConstant.IZ_NOTIFICATION_SOUND:
                try {
                    String soundName = (String) call.arguments;
                    DATB.setNotificationSound(soundName);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case MoMagicConstant.IZ_ADD_TAGS:
                try {
                    List<String> addTagList = new ArrayList<>();
                    addTagList = (List<String>) call.arguments;
                    DATB.addTag(addTagList);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case MoMagicConstant.IZ_REMOVE_TAG:
                try {
                    List<String> addTagList = new ArrayList<>();
                    addTagList = (List<String>) call.arguments;
                    DATB.removeTag(addTagList);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case iZootoConstant.IZ_DEFAULT_TEMPLATE:
                try {
                    int notificationTemplate = call.argument(MoMagicConstant.IZ_DEFAULT_TEMPLATE);
                   // setCustomNotification(notificationTemplate);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case iZootoConstant.IZ_DEFAULT_NOTIFICATION_BANNER:
//                try {
//                    String notificationTemplateBanner = call.argument(MoMagicConstant.IZ_DEFAULT_NOTIFICATION_BANNER);
//                    if (getBadgeIcon(context, notificationTemplateBanner) != 0) {
//                        DATB.setDefaultNotificationBanner(getBadgeIcon(context, notificationTemplateBanner));
//                    }
//                } catch (Exception ex) {
//                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
//                }
                break;

            case iZootoConstant.IZ_HANDLE_NOTIFICATION:
                try {
                    Object handleNotification = call.argument(MoMagicConstant.IZ_HANDLE_NOTIFICATION);
                    Map<String, String> map = (Map<String, String>) handleNotification;
                    if (map != null && !map.isEmpty()) {
                        //  DATB.iZootoHandleNotification(context, map);
                    }
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case iZootoConstant.IZ_RECEIVED_PAYLOAD:
                try {
                   // iZootoNotificationListener.onNotificationReceivedHybrid(notificationPayload);
                    //DATB.notificationReceivedCallback(iZootoNotificationListener);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case iZootoConstant.IZ_OPEN_NOTIFICATION:
                try {
                  //  iZootoNotificationListener.onNotificationOpened(notificationOpenedData);
                   // iZooto.notificationClick(iZootoNotificationListener);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            case iZootoConstant.IZ_DEVICE_TOKEN:
                try {

                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;




            /*   navigateToSettings    */
            case iZootoConstant.IZ_NAVIGATE_SETTING:
                try {
                    DATB.navigateToSettings(activity);
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;

            // Android 13 - Notification permission
            case iZootoConstant.IZ_NOTIFICATION_PERMISSION:
                try {
                    Log.e("Permission","Calling");
                    if (Build.VERSION.SDK_INT >= 33) {
                        DATB.promptForPushNotifications();
                    } else {
                        Log.e(MoMagicConstant.IZ_NOTIFICATION_PERMISSION, iZootoConstant.IZ_API_LEVEL_ERROR);
                    }
                } catch (Exception ex) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString());
                }
                break;








            default:
                result.notImplemented();
                break;
        }
    }


    private void runOnMainThread(final Runnable runnable) {
        if (Looper.getMainLooper().getThread() == Thread.currentThread())
            runnable.run();
        else {
            Handler handler = new Handler(Looper.getMainLooper());
            handler.post(runnable);
        }
    }






}

File: momagic_flutter/android/src/main/kotlin/com/example/momagic_flutter/MomagicFlutterPlugin.kt
package com.example.momagic_flutter

import android.app.Activity
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.gson.Gson
import com.momagic.AppConstant
import com.momagic.DATB
import com.momagic.NotificationHelperListener
import com.momagic.NotificationReceiveHybridListener
import com.momagic.NotificationWebViewListener
import com.momagic.Payload
import com.momagic.PreferenceUtil
import com.momagic.PushTemplate
import com.momagic.TokenReceivedListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONArray


/** MomagicFlutterPlugin */
class MomagicFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var activity: Activity? = null
    private var notificationOpenedData: String? = null
    private var notificationToken: String? = null
    private var notificationWebView: String? = null
    private var notificationPayload: String? = null

    companion object {
        private const val TAG = "MOMAGIC_FLUTTER"
    }

    // ActivityAware methods
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this.activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        this.activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        this.activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        this.activity = null
    }

    // Flutter plugin binding
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        try {
            context = flutterPluginBinding.applicationContext
            channel =
                MethodChannel(flutterPluginBinding.binaryMessenger, MoMagicConstant.IZ_PLUGIN_NAME)
            channel.setMethodCallHandler(this)
        } catch (ex: Exception) {
            Log.d(TAG, ex.toString())
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        try {
            channel.setMethodCallHandler(null)
        } catch (ex: Exception) {
            Log.d(TAG, ex.toString())
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            val moMagicNotificationListener = MoMagicNotificationListener()
            val preferenceUtil = PreferenceUtil.getInstance(context)

            when (call.method) {

                MoMagicConstant.IZ_ANDROID_INIT -> {
                    try {
                        DATB.isHybrid = true
                        val isDefaultWebView =
                            call.argument<Boolean>(MoMagicConstant.IZ_DEFAULT_WEB_VIEW) ?: false
                        preferenceUtil.setBooleanData(
                            AppConstant.IZ_DEFAULT_WEB_VIEW,
                            isDefaultWebView
                        )

                        if (isDefaultWebView) {
                            DATB.initialize(context)
                                .setTokenReceivedListener(moMagicNotificationListener)
                                .setNotificationReceiveListener(moMagicNotificationListener)
                                .setNotificationReceiveHybridListener(moMagicNotificationListener)
                                .build()
                        } else {
                            DATB.initialize(context)
                                .setTokenReceivedListener(moMagicNotificationListener)
                                .setNotificationReceiveListener(moMagicNotificationListener)
                                .setNotificationReceiveHybridListener(moMagicNotificationListener)
                                .setLandingURLListener(moMagicNotificationListener)
                                .build()
                        }

                        DATB.setPluginVersion(MoMagicConstant.IZ_PLUGIN_VERSION)
                    } catch (ex: Exception) {
                        Log.e(TAG, ex.toString())
                    }
                }

                MoMagicConstant.IZ_SET_SUBSCRIPTION -> {
                    try {
                        val setSubscription = call.arguments as Boolean
                        DATB.setSubscription(setSubscription)
                    } catch (ex: Exception) {
                        Log.e(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_PLUGIN_SUBSCRIBER_ID -> {
                    try {
                        val subscriberId = call.arguments as String
                        DATB.setSubscriberID(subscriberId)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_FIREBASE_ANALYTICS -> {
                    try {
                        val trackFirebaseAnalytics = call.arguments as Boolean
                        DATB.setFirebaseAnalytics(trackFirebaseAnalytics)
                    } catch (ex: Exception) {
                        Log.e(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_ADD_EVENTS -> {
                    try {
                        val eventName = call.argument<String>(MoMagicConstant.IZ_EVENT_NAME)
                        val hashMapEvent =
                            call.argument<HashMap<String, Any>>(MoMagicConstant.IZ_EVENT_VALUE)
                        DATB.addEvent(eventName, hashMapEvent)
                    } catch (ex: Exception) {
                        Log.e(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

               MoMagicConstant.IZ_ADD_PROPERTIES -> {
                    try {
                        val hashMapUserProperty = call.arguments as? HashMap<String, Any>
                        if (hashMapUserProperty != null) {
                            DATB.addUserProperty(hashMapUserProperty)
                        } else {
                            Log.e("UserProperties", "Received null map")
                        }
                    } catch (ex: Exception) {
                        Log.e("UserProperties", "Exception", ex)
                    }
                }






                MoMagicConstant.IZ_NOTIFICATION_SOUND -> {
                    try {
                        val soundName = call.arguments as String
                        DATB.setNotificationSound(soundName)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_ADD_TAGS -> {
                    try {
                        val addTagList = call.arguments<List<String>>()
                        DATB.addTag(addTagList)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_REMOVE_TAG -> {
                    try {
                        val addTagList = call.arguments<List<String>>()
                        DATB.removeTag(addTagList)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_DEFAULT_TEMPLATE -> {
                    try {
                        val notificationTemplate =
                            call.argument<Int>(MoMagicConstant.IZ_DEFAULT_TEMPLATE) ?: 0
                        setCustomNotification(notificationTemplate)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_DEFAULT_NOTIFICATION_BANNER -> {
                    try {
                        val notificationTemplateBanner =
                            call.argument<String>(MoMagicConstant.IZ_DEFAULT_NOTIFICATION_BANNER)
                        notificationTemplateBanner?.let {
                            if (getBadgeIcon(context, it) != 0) {
                                DATB.setDefaultNotificationBanner(getBadgeIcon(context, it))
                            }
                        }
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_RECEIVED_PAYLOAD -> {
                    try {
                        moMagicNotificationListener.onNotificationReceivedHybrid(notificationPayload)
                        DATB.notificationReceivedCallback(moMagicNotificationListener)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_OPEN_NOTIFICATION -> {
                    try {
                        moMagicNotificationListener.onNotificationOpened(notificationOpenedData)
                        DATB.notificationClick(moMagicNotificationListener)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_DEVICE_TOKEN -> {
                    try {
                        moMagicNotificationListener.onTokenReceived(notificationToken)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_HANDLE_WEB_VIEW -> {
                    if (!preferenceUtil.getBoolean(AppConstant.IZ_DEFAULT_WEB_VIEW)) {
                        try {
                            moMagicNotificationListener.onWebView(notificationWebView)
                            DATB.notificationWebView(moMagicNotificationListener)
                        } catch (ex: Exception) {
                            Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                        }
                    }
                }

                MoMagicConstant.IZ_CHANNEL_NAME -> {
                    try {
                        val channelName = call.arguments as String
                        DATB.setNotificationChannelName(channelName)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_NAVIGATE_SETTING -> {
                    try {
                        DATB.navigateToSettings(activity)
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                MoMagicConstant.IZ_NOTIFICATION_PERMISSION -> {
                    try {
                        if (Build.VERSION.SDK_INT >= 33) {
                            DATB.promptForPushNotifications()
                        } else {
                            Log.d(
                                MoMagicConstant.IZ_PLUGIN_EXCEPTION,
                                MoMagicConstant.IZ_API_LEVEL_ERROR
                            )
                        }
                    } catch (ex: Exception) {
                        Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
                    }
                }

                else -> {
                    result.notImplemented()
                }
            }

        } catch (ex: Exception) {
            Log.e(TAG, ex.message.toString())
        }
    }

    private fun runOnMainThread(runnable: Runnable) {
        try {
            if (Looper.getMainLooper().thread == Thread.currentThread()) {
                runnable.run()
            } else {
                Handler(Looper.getMainLooper()).post(runnable)
            }
        } catch (ex: Exception) {
            Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
        }
    }

    fun invokeMethodOnUiThread(methodName: String, map: String) {
        try {
            runOnMainThread {
                channel.invokeMethod(methodName, map)
            }
        } catch (ex: Exception) {
            Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
        }
    }

    private fun setCustomNotification(index: Int) {
        try {
            when (index) {
                2 -> DATB.setDefaultTemplate(PushTemplate.TEXT_OVERLAY)
                3 -> DATB.setDefaultTemplate(PushTemplate.DEVICE_NOTIFICATION_OVERLAY)
                else -> DATB.setDefaultTemplate(PushTemplate.DEFAULT)
            }
        } catch (ex: Exception) {
            Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, ex.toString())
        }
    }

    private fun getBadgeIcon(context: Context, setBadgeIcon: String): Int {
        var badgeIcon = 0
        try {
            val drawableId =
                context.resources.getIdentifier(setBadgeIcon, "drawable", context.packageName)
            if (drawableId != 0) {
                badgeIcon = drawableId
            } else {
                val mipmapId =
                    context.resources.getIdentifier(setBadgeIcon, "mipmap", context.packageName)
                if (mipmapId != 0) {
                    badgeIcon = mipmapId
                }
            }
        } catch (e: Exception) {
            Log.v(AppConstant.APP_NAME_TAG, e.toString())
        }
        return badgeIcon
    }


    // MoMagic Listeners Implementation
    private inner class MoMagicNotificationListener : NotificationHelperListener,
        NotificationWebViewListener, TokenReceivedListener, NotificationReceiveHybridListener {

        override fun onNotificationReceived(payload: Payload?) {
            if (payload != null) {
                try {
                    val jsonPayload = Gson().toJson(payload)
                    invokeMethodOnUiThread(MoMagicConstant.IZ_RECEIVED_PAYLOAD, jsonPayload)
                } catch (e: Exception) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, e.toString())
                }
            }
        }

        override fun onNotificationOpened(data: String?) {
            notificationOpenedData = data
            if (data != null) {
                try {
                    invokeMethodOnUiThread(MoMagicConstant.IZ_OPEN_NOTIFICATION, data)
                } catch (e: Exception) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, e.toString())
                }
            }
        }

        override fun onWebView(landingUrl: String?) {
            notificationWebView = landingUrl
            if (landingUrl != null) {
                try {
                    invokeMethodOnUiThread(MoMagicConstant.IZ_HANDLE_WEB_VIEW, landingUrl)
                } catch (e: java.lang.Exception) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, e.toString())
                }
            }
        }

        override fun onTokenReceived(token: String?) {
            notificationToken = token
            if (token != null) {
                try {
                    invokeMethodOnUiThread(MoMagicConstant.IZ_DEVICE_TOKEN, token)
                } catch (e: java.lang.Exception) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, e.toString())
                }
            }
        }

        override fun onNotificationReceivedHybrid(data: String?) {
            notificationPayload = data
            if (data != null) {
                try {
                    val listArray = JSONArray(data)
                    val reverseList = JSONArray()
                    if (listArray.length() > 0) {
                        reverseList.put(listArray.getJSONObject(listArray.length() - 1))
                    }
                    invokeMethodOnUiThread(
                        MoMagicConstant.IZ_RECEIVED_PAYLOAD,
                        reverseList.toString()
                    )
                } catch (e: java.lang.Exception) {
                    Log.v(MoMagicConstant.IZ_PLUGIN_EXCEPTION, e.toString())
                }
            }
        }
    }

}
